### PR TITLE
[Security] Add authentication success sensitive event to authentication provider manager (rebase #26050)

### PR DESCRIPTION
**Pull request by curry684**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #26050 #18494
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/11334

Rebased #26050 at EUFOSSA 

Thanks @robfrawley for the excellent work!

### CODE REVIEW

**Comment by curry684** [comment]
@wouterj can you pick up the doc on this one?

File: src/Symfony/Component/Security/Core/Event/AuthenticationSensitiveEvent.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Core\Event;
+
+use Symfony\Component\EventDispatcher\Event;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Core/Event/AuthenticationSensitiveEvent.php]: ` * @deprecated since Symfony 4.3, use "Symfony\Contracts\EventDispatcher\Event" instead`

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Security/Core/Event/AuthenticationSensitiveEvent.php (hunk 1 above)
Comment (curry684) [reply src/Symfony/Component/Security/Core/Event/AuthenticationSensitiveEvent.php]: Shoot, fixed tons of those during the merge but missed this one hehe.

Comment (curry684) [reply src/Symfony/Component/Security/Core/Event/AuthenticationSensitiveEvent.php]: Mmmm and I see why, the entire Security component wasn't done yet, see https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Security/Core/Event/AuthenticationEvent.php

Comment (curry684) [reply src/Symfony/Component/Security/Core/Event/AuthenticationSensitiveEvent.php]: Done, and I still hate you   :wink:

File: src/Symfony/Component/Security/Core/Event/AuthenticationEvent.php [hunk 2]
```diff
@@ -11,8 +11,8 @@
 
 namespace Symfony\Component\Security\Core\Event;
 
-use Symfony\Component\EventDispatcher\Event;
 use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Contracts\EventDispatcher\Event;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Core/Event/AuthenticationEvent.php]: should be reverted: BC break :)

File: src/Symfony/Component/Security/Core/Event/AuthenticationSensitiveEvent.php [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+ *
+ * @author Rob Frawley 2nd <rmf@src.run>
+ */
+class AuthenticationSensitiveEvent extends Event
+{
+    private $preAuthenticationToken;
+    private $authenticationToken;
+    private $authenticationProviderClassName;
+
+    public function __construct(TokenInterface $preAuthenticationToken, TokenInterface $authenticationToken, ?string $authenticationProviderClassName = null)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Core/Event/AuthenticationSensitiveEvent.php]: please remove the`?` as our CS is to consider the `null` default means the same (it does)

File: src/Symfony/Component/Security/Core/Event/AuthenticationSensitiveEvent.php [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+     *
+     * @param \Closure|null $extractor An optional custom token credentials password extraction \Closure that is
+     *                                 provided an auth token (as an instance of TokenInterface) and an auth event
+     *                                 (as an instance of AuthenticationSensitiveEvent). This closure is called
+     *                                 first with the final-auth token and second with the pre-auth token, returning
+     *                                 early if a non-null/non-empty scalar/castable-object value is returned.
+     *
+     * @return string|null Either a credentials password/secret/auth_key is returned or null on extraction failure
+     */
+    public function getAuthenticationTokenPassword(?\Closure $extractor = null): ?string
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Core/Event/AuthenticationSensitiveEvent.php]: ~`?`~

File: src/Symfony/Component/Security/Core/Event/AuthenticationSensitiveEvent.php [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                ?: $this->tryArrayFindCredentialsPasswordToString($credentials);
+        };
+
+        return ($extractor($this->authenticationToken, $this) ?: null)
+            ?: ($extractor($this->preAuthenticationToken, $this) ?: null);
+    }
+
+    private function tryCoercibleCredentialsPasswordToString($credentials): ?string
+    {
+        return is_scalar($credentials) || method_exists($credentials, '__toString')
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Core/Event/AuthenticationSensitiveEvent.php]: should be on one line as we usually prefer long line in similar cases

File: src/Symfony/Component/Security/Core/Tests/Event/AuthenticationSensitiveEventTest.php [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+                if (isset($c[$k = sprintf('%s-key', $default)])) {
+                    return $c[$k];
+                }
+
+                return $default;
+            };
+        };
+
+        $createHasStrCastableClass = function (?string $return = null) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Core/Tests/Event/AuthenticationSensitiveEventTest.php]: ~`?`~ (same below)

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]
Now, you can hate me ;)

**Comment by curry684** [comment]
Get me my headset back and I'll sort the mess out again :smile:

**Comment by javiereguiluz** [comment]
Event names are usually very descriptive ... so I find this new event name (`AUTHENTICATION_SUCCESS_SENSITIVE`) really unfortunate. It doesn't tell me at all what's this about :(  All security events are "sensitive".

**Comment by curry684** [comment]
@javiereguiluz fair point but I don't really see an alternative. Technically it IS the authentication success event, just called a second time before we do `eraseCredentials` so it's a sensitive variant of the success event. We could do something like `AUTHENTICATION_SUCCESS_WITH_TOKEN` or something but that seems overly verbose and no longer readable.

I'll take any suggestions here...

**Comment by aschempp** [comment]
Isn't Symfony getting rid of event names anyway?
https://symfony.com/blog/new-in-symfony-4-3-simpler-event-dispatching

**Comment by curry684** [comment]
@aschempp correct but it would be inconsistent right now to follow that convention for the new event, and major B/C breakage to adapt all the others. So in core that's a job for later on.

**Comment by fabpot** [comment]
I agree that we cannot remove the old ones, but there is no need to create a new one. People will eventually remove usage of the current names, so the inconsistency is not a problem IMHO.

**Comment by curry684** [comment]
Fair enough, and just thinking some more about the name => how about `AuthenticationAcceptedEvent`?

It conveys that we do not yet consider the Authentication a success (because if you throw inside the Accepted it will fail still) but that we can confirm we are about to, and as such this is the time to do last call on the token that was accepted before we clean it up.

**Comment by aschempp** [comment]
Maybe `AuthenticationCredentialsEvent` as the event is only there to get the credentials?

**Comment by curry684** [comment]
As credentials are part of the token it would then have to be `AuthenticationTokenEvent`, but generally in Events we follow (rightfully imo) the convention of naming it after what happens instead of what is communicated (the payload of the event). Hence my suggestion as the event is dispatched at the point where we have accepted the token.

**Comment by nicolas-grekas** [comment]
> Isn't Symfony getting rid of event names anyway?

I'm not sure we are: core should be much more explicit than what is allowed for not-core. E.g. autowiring is not a feature we use in core.

**Comment by nicolas-grekas** [comment]
oh, but the new event should be listed in `src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml`

**Comment by nicolas-grekas** [comment]
Thinking a bit about this approach, I have a different one to propose, see #31139

**Comment by javiereguiluz** [comment]
@wouterj do you think this proposal fits well with the latest changes you are proposing in the Security component or is it unrelated? Thanks!

**Comment by wouterj** [comment]
It's a bit hard to follow, as the discussion was inside multiple PRs and issues (#26050, #31139, #31843, #18494), but as far as I can see this PR is no longer needed.

#31843 build another approach to rehash passwords that was merged into 4.3 (https://symfony.com/blog/new-in-symfony-4-3-native-password-encoder). As far as I can see, rehashing credentials was the only reason for this event to exists.

**Comment by chalasr** [comment]
As said, password rehashing is now supported. I'm going to close here until another use case comes out. Thank you @robfrawley and @curry684
